### DISCUSSION TITLE
majority element

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/progettazione-di-algoritmi#54] [Esercizi A-L]
Dato un array $A$, un majority element in un array è un valore $x$ tale che almeno $\frac{n}{2}$ elementi dell'array sono uguali a $x$.

Scrivere lo pseudocodice di un algoritmo che decide se esiste un majority element in un array in $O(n log(n))$

**Comment by 18xChris** [comment]
def solve(self,nums,i,j):
        if i == j:
            return nums[i]
        m = (i+j)//2
        sx = self.solve(nums,i,m)
        dx = self.solve(nums,m+1,j)


        if sx == dx:
            print("entra")
            return sx
        if sx == None:
            return dx
        if dx == None:
            return sx
        csx = 0  
        cdx = 0
        for k in range(i,j+1):
            if nums[k]  != None:
                if nums[k] == sx:
                    csx+=1
                if nums[k] == dx:
                    cdx+=1
        if csx > cdx:
            return sx
        if csx < cdx:
            return dx
            
        return None

**Comment by Alegau03** [comment]
Def ElemntoMaggioritario(A):
    if len(A)==1:
        return A[0]
    m= A.len()/2
    m1 = EleemntoMaggioritario(A[0:m])
    m2 = EleemntoMaggioritario(A[m+1::])

    count = 0

    for i in A:
        if A[i]==m1:
            count+=1
    
    if count>A.len()/2:
        return m1

    count = 0

    for i in A:
        if A[i]==m2:
            count+=1
    
    if count>A.len()/2:
        return m2

**Comment by daromacarlo** [comment]
```c
def maj(A: array ordinato di cardinalità dispari)
	if len(A) == 0{
		return Null
	}
	if len(A) == 1{
		return A
	}
	if len(A) == 2 && A[0] == A[1]{
		return A[0]					// oppure A[1]
	}
	m = len(A) / 2
	R = maj(A[0:m])
	L = maj(A[m+1:len(A)-1])
	count = 0

	for i=0,i<=len(A)-1,i++{
		if A[i] == R
		count ++
	}

	if count > m{
		return R
	}

	count = 0
	for i=0,i<=len(A)-1,i++{
		if A[i] == R
		count ++
	}

	if count > m{
		return L
	}

	return Null
	

```
